### PR TITLE
feat(throttle+retry): support octokit plugins

### PR DESCRIPTION
**Pull request by DeviaVir**
Related to https://github.com/zaproxy/action-full-scan/issues/59

I thought this would be a simple change, but it turned out that the `octokit` instance the plugins were compatible with, required ESM, and could not be used with `rest` at all, which means I had to rewrite much more than I wanted to.

I will test this in my own environments first, but wanted to contribute for early review.

For testing purposes:

* https://github.com/DeviaVir/action-full-scan
* https://github.com/DeviaVir/action-baseline

Cheers.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -185,13 +204,13 @@ const actionCommon = {
 
     const runnerInfo = `RunnerID:${currentRunnerID}`;
     const runnerLink =
-      `View the [following link](${context.serverUrl}/${owner}/${repo}/actions/runs/${currentRunnerID})` +
+      `View the [following link](https://github.com/${owner}/${repo}/actions/runs/${currentRunnerID})` +
```
Comment (thc202) [inline src/index.ts:207]: Isn't this going to break GHES users?

File: src/index.ts [hunk 2]
```diff
@@ -33,7 +35,7 @@ const actionCommon = {
     let openIssue:
       | components["schemas"]["issue-search-result-item"]
       | undefined;
-    let currentReport: Report | undefined;
+    let currentReport: Report | undefined
```
Comment (thc202) [inline src/index.ts:38]: Why these changes?

File: src/index.ts (hunk 1 above)
Comment (DeviaVir) [reply src/index.ts:207]: Yes, but the problem is we no longer have a `context.serverUrl`, and I didn't immediately find a great way to replace this

File: src/index.ts (hunk 2 above)
Comment (DeviaVir) [reply src/index.ts:38]: Not on purpose, code linter

File: src/index.ts (hunk 1 above)
Comment (DeviaVir) [reply src/index.ts]: OK, I found out `process.env.GITHUB_SERVER_URL` holds this information, updated.

File: src/index.ts (hunk 2 above)
Comment (DeviaVir) [reply src/index.ts]: Reverted.

**Comment by thc202** [comment]
Build is failing.

File: package.json [hunk 3]
```diff
@@ -1,10 +1,14 @@
 {
   "name": "@zaproxy/actions-common-scans",
-  "version": "0.2.0",
+  "version": "0.2.1",
```
Comment (thc202) [inline package.json:3]: This should be reverted, the version is managed automatically.

File: src/action-helper.ts [hunk 4]
```diff
@@ -232,15 +232,15 @@ const actionHelper = {
 
         console.log(
           `#${newAlerts.length} alerts have been identified` +
-            ` and #${removedAlerts.length} alerts have been ignored for the site.`,
+          ` and #${removedAlerts.length} alerts have been ignored for the site.`,
```
[the remark this answers is not in the document]
Comment (DeviaVir) [reply src/action-helper.ts]: Interesting clash between linter and eslint, but yes, resolved

**Comment by DeviaVir** [comment]
> Build is failing:
> 
> > eslint src/
> > Oops! Something went wrong! :(
> > ESLint: 8.57.0
> > Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/runner/work/actions-common/actions-common/.eslintrc.js from /Users/runner/work/actions-common/actions-common/node_modules/@eslint/eslintrc/dist/eslintrc.cjs not supported.
> > .eslintrc.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
> > Instead either rename .eslintrc.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /Users/runner/work/actions-common/actions-common/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

Got it, should be fixed now.

**Comment by thc202** [comment]
Some tests failing, might need mock adjustments?

> > @zaproxy/actions-common-scans@0.2.0 test
> > jest
> 
> PASS src/models/FilteredSite.test.ts
> FAIL src/index.test.ts (6.394 s)
>   ● processReport › with a report file in place › without an open issue › creates a new issue
> 
>     HttpError: Bad credentials - https://docs.github.com/rest
> 
>       at node_modules/@octokit/request/dist-bundle/index.js:106:21
>       at requestWithGraphqlErrorHandling (node_modules/@octokit/plugin-retry/dist-bundle/index.js:36:20)
>       at Job.doExecute (node_modules/bottleneck/light.js:405:18)
> 
> 
> Error: HttpError: Bad credentials - https://docs.github.com/rest
> 
>     at node_modules/@octokit/request/dist-bundle/index.js:106:21
>     at requestWithGraphqlErrorHandling (node_modules/@octokit/plugin-retry/dist-bundle/index.js:36:20)
>     at Job.doExecute (node_modules/bottleneck/light.js:405:18)
> 
> Errors thrown in src/index.test.ts
> Test Suites: 1 failed, 1 passed, 2 total
> Tests:       1 failed, 1 passed, 2 total
> Snapshots:   0 total
> Time:        6.695 s
> Ran all test suites.
> Error: Process completed with exit code 1.

**Comment by DeviaVir** [comment]
I had to switch to fetch-mock package instead, please take a fresh look, the tests are passing now. @thc202

**Comment by tjcampanella** [comment]
@thc202 Can this be merged? We run into the secondary rate limit issue a lot on a couple of our repos.

**Comment by psiinon** [comment]
This PR now has conflicts :/

File: package.json [hunk 5]
```diff
@@ -53,10 +66,11 @@
...
     "eslint-plugin-prettier": "^5.1.3",
-    "jest": "^29.2.2",
+    "fetch-mock": "^10.1.1",
+    "jest": "^29.7.0",
     "prettier": "^3.2.5",
     "semantic-release": "^24.0.0",
-    "ts-jest": "^29.0.3",
+    "ts-jest": "^29.1.2",
     "typescript": "^5.0.4",
     "undici": "^6.7.1"
```
Comment (thc202) [inline package.json:75]: This one should no longer be needed per mock removals.

File: src/index.test.ts [hunk 6]
```diff
@@ -1,10 +1,9 @@
-import { main } from "./index";
+import { main } from "./index.js";
 import fs from "fs";
-import { Interceptable, MockAgent, setGlobalDispatcher } from "undici";
-import { ReportFixture } from "./models/ReportFixture";
-import { mockSearchIssues } from "./testHelpers/githubApiMocks/mockSearchIssues";
-import { mockCreateIssue } from "./testHelpers/githubApiMocks/mockCreateIssue";
```
Comment (thc202) [inline src/index.test.ts:6]: The mocks no longer seem to be needed.

File: package.json (hunk 5 above)
Comment (DeviaVir) [reply package.json]: patched

**Comment by DeviaVir** [comment]
Resolved conflicts and implemented feedback.

I don't mean to come across as mean, but I don't plan on maintaining this PR forever, so if there are only nit comments in the future please feel free to implement these, I plan on closing this PR when the next conflict arises.

**Comment by thc202** [comment]
That sounds good to me.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Comment by DeviaVir** [comment]
Thanks for reviewing and merging! 👊🏻
